cubi_tk/isa_tab/annotate.py

```diff
@@ -0,0 +1,457 @@
+
+        # Read isa-tab file
+        isa_data = isa_support.load_investigation(self.config.input_investigation_file)
+
+        # Check target study/assay availability
+        if not self._check_studies_and_assays(isa_data):
+            return 1
+
+        # Read annotation file and build mapping
+        annotation_data = self._read_annotation(self.config.input_annotation_file)
+        annotation_map, header_map = self._build_annotation_map(annotation_data)
+
+        # Add annotation
+        self._perform_update(isa_data, annotation_map, header_map)
+
+        return 0
+
+    def _check_studies_and_assays(self, isa_data):
+        # List studies
+        study_file_names = list(isa_data.studies.keys())
+        # Check that a least one study exists
+        if len(study_file_names) > 0:
+            # If no target study declared, use first study
+            if not self.config.target_study:
+                self.config.target_study = study_file_names[0]
+            # Check if target study is in list (i.e. in investigation)
+            elif self.config.target_study not in study_file_names:
+                logger.error(
+                    "Invalid target study '%s'.\nStudies available:\n%s",
+                    self.config.target_study,
+                    study_file_names,
+                )
+                return False
+        else:
+            logger.error("No studies available in investigation.")
+            return False
+
+        # List assays
+        assay_file_names = list(isa_data.assays.keys())
+        # Check that a least one assay exists
+        if len(assay_file_names) > 0:
+            # If no target assay declared, use first assay
+            if not self.config.target_assay:
+                self.config.target_assay = assay_file_names[0]
+            # Check if target assay is in list (i.e. in investigation)
+            if self.config.target_assay not in assay_file_names:
+                logger.error(
+                    "Invalid target assay '%s'.\nAssays available:\n%s",
+                    self.config.target_assay,
+                    assay_file_names,
+                )
+                return False
+        else:
+            logger.error("No assays available in investigation.")
+            return False
+
+        return True
+
+    def _read_annotation(self, filename):
+        with open(filename) as infile:
+            anno_reader = csv.reader(infile, delimiter="\t")
+            annotation_data = []
+            for row in anno_reader:
+                annotation_data.append(row)
+
+        # Need at least two columns (e.g. one source/sample name plus new annotation) and one data row
+        if len(annotation_data) < 2 and len(annotation_data[0]) < 2:
+            logger.error("Annotation file needs at least two columns and two rows")
+            return 1
+        return annotation_data
+
+    def _build_annotation_map(self, annotation):
+        # change to long df
```

Comment:
Format this comment a bit better, e.g., use `Change to long def.` (proper lower/upper case) with an empty line afterwards.